### PR TITLE
Resource and Component unification refactor

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: assets/system/src/lib.rs [hunk 1]
```diff
@@ -23,13 +23,19 @@ blueprint! {
         }
 
         /// Mints fungible resource. TODO: Remove
-        pub fn mint(amount: Decimal, resource_address: ResourceAddress) -> Bucket {
-            resource_manager!(resource_address).mint(amount)
+        pub fn mint(
+            amount: Decimal,
+            resource_address: ResourceAddress,
+            auth: Bucket,
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply assets/system/src/lib.rs:29]: That's right but this is something we can change/discuss about from product point of view as I think we haven't thought about mint/burn permission to components.

File: radix-engine/src/engine/track.rs [hunk 2]
```diff
@@ -445,6 +459,30 @@ impl<'s, S: SubstateStore> Track<'s, S> {
...
+            self.resource_managers.insert(
+                resource_address.clone(),
+                SubstateUpdate {
+                    prev_id: substate_update.prev_id,
+                    value: resource_manager,
+                },
+            );
+            Ok(result)
+        } else {
+            panic!("Resource Manager not found");
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/track.rs:482]: This is a temporary panic due to bad interface but will be fixed in a future PR. Would you prefer a comment regarding this?

File: radix-engine/src/engine/process.rs [hunk 3]
```diff
@@ -52,19 +54,37 @@ macro_rules! re_warn {
...
+    Bucket(Bucket),
+    Vault(VaultId),
+}
+
+#[derive(Debug, Clone)]
+pub enum SNodeRef {
+    Scrypto(ScryptoActor),
+    Resource(ResourceAddress),
+    Bucket(BucketId),
+    Vault(VaultId),
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/process.rs:72]: Just a way to address for invocation

File: radix-engine/src/engine/process.rs [hunk 4]
```diff
@@ -1786,131 +1828,96 @@ impl<'r, 'l, L: SubstateStore> Process<'r, 'l, L> {
...
-            .ok_or(RuntimeError::ResourceManagerNotFound(
-                input.resource_address,
-            ))?;
-        resource_manager
-            .update_metadata(input.new_metadata)
-            .map_err(RuntimeError::ResourceManagerError)?;
-
+        let invocation = Invocation {
+            snode_ref: SNodeRef::Resource(input.resource_address.clone()),
+            function: "update_metadata".to_string(),
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/process.rs:1842]: Possible but I'd like to leave as is as this problem is also the same problem on the Scrypto side. By unifying the interface we can come up with the correct solution to both.

**Comment by iamyulong** [comment]
Had a sync with Josh and agreed to
- Keep the modeling for now, as all Scrypto APIs will be kept intact
- Discuss about authorization implication on team meeting
- Not add compiler type checks on "system component" methods (refactoring is ongoing and we're trying to avoid churns), but should have that in the long run

**Review by iamyulong — APPROVED** [review]

**Pull request merged**